### ISSUE TITLE
why one SparseSequential cost much less time comparing with several SparseSequentials' combination?

### ISSUE THREAD

**Issue by bigsheep2012** [traveller59/spconv#28]
Hello Yan @traveller59,

As the title mentioned. 
The following two examples are with logically equal net structure with your spconv, but time costs are very different. With a specific point cloud (around 50000 points ), the entire time duration for the first example is 70~90ms, while the second example cost around 200 ms. The input shape is [1280,1280, 50], num_input_features = 4
1) 
self.middle_spconv = spconv.SparseSequential(
    spconv.SubMConv3d(num_input_features, 16, 3, bias=False, indice_key="subm0"),
    nn.BatchNorm1d(16, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(16, 16, 3, bias=False, indice_key="subm0"),
    nn.BatchNorm1d(16, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(16, 32, 3, 2, bias=False, padding=1), # [1280, 1280, 50] -> [640, 640, 25]
    nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(32, 32, 3, bias=False, indice_key="subm1"),
    nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(32, 32, 3, bias=False, indice_key="subm1"),
    nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(32, 64, 3, 2, bias=False, padding=1), # [640, 640, 25] -> [320, 320, 13]
    nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(64, 64, 3, bias=False, indice_key="subm2"),
    nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(64, 64, 3, bias=False, indice_key="subm2"),
    nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(64, 96, 3, 2, bias=False, padding=1),  # [320, 320, 13] -> [160, 160, 7]
    nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
    nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
    nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
    nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(96, 128, 3, 2, bias=False, padding=1),  # [160, 160, 7] -> [80, 80,4]
    nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
    nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
    nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
    nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(128, 160, 3, 2, bias=False, padding=1),  # [80, 80, 4] -> [40, 40, 2]
    nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
    nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
    nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
    nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
    nn.ReLU(),
    spconv.SparseConv3d(160, 192, 3, 2, bias=False, padding=1),  # [40, 40, 2] -> [20, 20, 1]
    nn.BatchNorm1d(192, eps=1e-3, momentum=0.01),
    nn.ReLU(),
)

2) 
` self.middle_conv_block1 = spconv.SparseSequential(
            spconv.SubMConv3d(num_input_features, 16, 3, bias=False, indice_key="subm0"),
            nn.BatchNorm1d(16, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(16, 16, 3, bias=False, indice_key="subm0"),
            nn.BatchNorm1d(16, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(16, 32, 3, 2, bias=False, padding=1), # [1280, 1280, 50] -> [640, 640, 25]
            nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )

        self.middle_conv_block2 = spconv.SparseSequential(
            spconv.SubMConv3d(32, 32, 3, bias=False, indice_key="subm1"),
            nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(32, 32, 3, bias=False, indice_key="subm1"),
            nn.BatchNorm1d(32, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(32, 64, 3, 2, bias=False, padding=1), # [640, 640, 25] -> [320, 320, 13]
            nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )

        self.middle_conv_block3 = spconv.SparseSequential(
            spconv.SubMConv3d(64, 64, 3, bias=False, indice_key="subm2"),
            nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(64, 64, 3, bias=False, indice_key="subm2"),
            nn.BatchNorm1d(64, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(64, 96, 3, 2, bias=False, padding=1),  # [320, 320, 13] -> [160, 160, 7]
            nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )

        self.middle_conv_block4 = spconv.SparseSequential(
            spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
            nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
            nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(96, 96, 3, bias=False, indice_key="subm3"),
            nn.BatchNorm1d(96, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(96, 128, 3, 2, bias=False, padding=1),  # [160, 160, 7] -> [80, 80,4]
            nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )

        self.middle_conv_block5 = spconv.SparseSequential(
            spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
            nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
            nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(128, 128, 3, bias=False, indice_key="subm4"),
            nn.BatchNorm1d(128, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(128, 160, 3, 2, bias=False, padding=1),  # [80, 80, 4] -> [40, 40, 2]
            nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )

        self.middle_conv_block6 = spconv.SparseSequential(
            spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
            nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
            nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SubMConv3d(160, 160, 3, bias=False, indice_key="subm5"),
            nn.BatchNorm1d(160, eps=1e-3, momentum=0.01),
            nn.ReLU(),
            spconv.SparseConv3d(160, 192, 3, 2, bias=False, padding=1),  # [40, 40, 2] -> [20, 20, 1]
            nn.BatchNorm1d(192, eps=1e-3, momentum=0.01),
            nn.ReLU(),
        )`
Specifically, the middle_conv_block2 cost around 100ms on my machine. Am I doing anything wrong or it is caused by spconv's mechanism?

Thanks in advance.

**Comment by bigsheep2012** [comment]
Update 
I remove all .dense() in-betweens in forward function and the speed becomes normal.
So if I transfer a sparse tensor to dense like the following example, the time cost of middle_conv_block2 would be increased. The dense() return an inference of ret1 and turn ret1 into dense format inplace, which lead to an dense input for next sparse convolution? 

    def forward(self, voxel_features, coors, batch_size):
        coors = coors.int()
        ret0 = spconv.SparseConvTensor(voxel_features, coors, self.sparse_shape, batch_size)

        ret1 = self.middle_conv_block1(ret0)
        ret1_dense = ret1.dense()
        
        ret2 = self.middle_conv_block2(ret1)
     return ret2

**Comment by traveller59** [comment]
```dense``` isn't a inplace operation, it convert sparse tensor to regular torch dense tensors without modify origin sparse tensor, this operation use pure pytorch implementation and may cost lots of time, check spconv/__init__.py for more details.
So the time cost increase shouldn't happen. How do you measure time? do you add ```torch.cuda.synchronize()```?

**Comment by bigsheep2012** [comment]
Nope. But I test the time outside the forward function. 
Like 
start = time.time()
out=model(input)
end=time.time()
print end-start
torch.cuda.synchronize() would affect this duration a lot?
Anyway, the time cost becomes much less after I remove all dense() operations. I guess the dense operations is time costly.

**Comment by traveller59** [comment]
you must use torch.cuda.synchronize() to measure device code time:
```Python
torch.cuda.synchronize()
t = time.time()
# device code
torch.cuda.synchronize()
print(time.time() - t)
```
otherwise the measured time is wrong.

**Comment by bigsheep2012** [comment]
I followed your instructions and got the same results.
That is, with .dense() operations in-betweens, the time cost is around 200ms:
0.196189165115
0.204177856445
while, without .dense() operations in-betweens, the time cost is around 70-90ms:
0.0651390552521
0.0862791538239

The example code is as follows:
def forward(self, voxel_features, coors, batch_size):
        coors = coors.int()
        ret0 = spconv.SparseConvTensor(voxel_features, coors, self.sparse_shape, batch_size)
        # ret0_dense = ret0.dense()
        torch.cuda.synchronize()
        t = time.time()
        
        ret1 = self.middle_conv_block1(ret0)  ## SparseSequential
        ret1_dense = ret1.dense()
        ret2 = self.middle_conv_block2(ret1) ## SparseSequential
        ret2_dense = ret2.dense()
        ret3 = self.middle_conv_block3(ret2) ## SparseSequential
        ret3_dense = ret3.dense()
        ret4 = self.middle_conv_block4(ret3) ## SparseSequential
        ret4_dense = ret4.dense()
        ret5 = self.middle_conv_block5(ret4) ## SparseSequential
        ret5_dense = ret5.dense()
        ret6 = self.middle_conv_block6(ret5) ## SparseSequential

        print time.time() - t
        return ret6.dense()
